[NVFuser] Upstream push 0907

jjsjann123 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #84626

Syncing nvfuser devel branch to upstream master. https://github.com/csarofeen/pytorch/

Codegen changes include:

- codegen improvement:
i. improved view support on pointwise and transpose scheduler
ii. grouped grid welford added for better outer-norm grid persistence in normalization

- misc:
i. new composite ops added: variance_mean , arange, 
ii. fixes misaligned address for transpose scheduler
iii. refactor on separation of compilation API from execution API to prepare us for async compilation
iv. double type support on expression evaluator
v. PYTORCH_NVFUSER_DUMP refactor to save PTX and CUBIN

Commits that's in this PR from the devel branch:
```
89330aa23aa804340b2406ab58899d816e3dc3d2 Tensor factories must set the output shape as its input (#1939)
b2fd01ea9346712c6d6f623ca6addbc4888d008e arange support (#1933)
56c00fd3922dad7dfc57351ad7d780f0f2f8e4ed Double support on all expression evaluators (#1937)
371f28223e57fe3f6b5e50a0a45177e6a5c0785c Improve trivial reduction merge support (#1931)
1d0c26790e5647920b40d419d26815bbe310b3a6 Test `rand` in a fusion with zero tensor input (#1932)
0dab160fb2177d178eef3148c6a529e0855009e9 Fix softmax bwd sizes. (#1890)
ef98f360f6d3e3e1cc662ecb65202d88150f128d Fix a bug (#1936)
63132a0c56508c550084b07fb76a3df865102d00 Propagate permissive mapping information into indexing pass (#1929)
b4ac2c88d78078ee4d8b21c4fc51645b5710a282 Map IterationDomains through view operations. (#1919)
c0a187a7619d7cf9dc920294e15461791e8d6d4d do not use deprecated functions (#1935)
88de85e758c5e4afb7b6e746573c0d9a53b4cea7 Upstream cherry pick fixes 0811 (#1934)
b247dcf7c57dc6ac3f7a799b0a6beb7770536a74 Separate kernel compilation API from kernel execution API (#1914)
b34e3b93ee1a8030730c14af3995dd95665af07d Fix `ir_utils::hasBlockSync` + misc fixes in transpose scheduler (#1924)
14a53e6707f43bf760494c238a46386d69830822 Nullary RNGOp (#1892)
3c3c89e638f5172cafb0761f22bacd1fd695eec3 Misc fixes/tuning for transpose scheduler (#1912)
20cf109c8b44d48f61977e35bae94368985144ac Grouped grid welford (#1921)
6cf7eb024c9e53c358cbe56597e117bad56efefd Transpose scheduler small dim sizes better support (#1910)
9341ea9a5bf42f9b14ccad0c94edbc79fc5bb552 Disabled ViewPersistentShmoo sizes that results in NAN (#1922)
057237f66deeea816bb943d802a97c1b7e4414ab Fix CUDA driver error: misaligned address for transpose scheduler  (#1918)
3fb3d80339e4f794767a53eb8fdd61e64cf404a2 Add variance_mean function using Welford (#1907)
98febf6aa3b8c6fe4fdfb2864cda9e5d30089262 Remove DisableOption::UnrollWithRng (#1913)
ee8ef33a5591b534cf587d347af11e48ba7a15d4 Minor fix for the debug interface of using PTX directly (#1917)
6e8f953351f9dabfd1f991d8431cecb6c2ce684d Add PYTORCH_NVFUSER_DUMP options to save PTX and CUBIN (#1916)
5eefa9a72385f6a4b145680a9dcc52d7e8293763 dopt is only available since nvrtc 11.7 (#1915)
2ec8fc711eafc72451eebf0f5e2a98a38bf3f6ef Kill computeAtBetween (#1911)
d0d106a1d9af118d71673173674e875be35d259d Improve view support on pointwise and transpose scheduler (#1906)
e71e1ecefe67219846070590bbed54bbc7416b79 Fix name clash of RNG with shared memory (#1904)
3381793a253689abf224febc73fd3fe2a0dbc921 Fix mutator and sameAs for expanded IterDomain (#1902)
```

RUN_TORCHBENCH: nvfuser

Differential Revision: [D39324552](https://our.internmc.facebook.com/intern/diff/D39324552)

jjsjann123 committed:
[NVFuser] Upstream push 0907

Syncing nvfuser devel branch to upstream master. https://github.com/csarofeen/pytorch/
Rebase and squashed commits from nvfuser upstream push 0907

RUN_TORCHBENCH: nvfuser

[ghstack-poisoned]

jjsjann123 commented:
:cry: forgot to add `trunk` label last night....

jjsjann123 commented:
errr. I'm seeing real build errors on macos and rocm test failures from the smoke test, which has the trunk label #84240 . start digging there...

jjsjann123 commented:
Rocm failure is not very informative for debugging. We are hitting a driver error 200 vvv:
```
2022-09-07T09:45:48.6209995Z RuntimeError: The following operation failed in the TorchScript interpreter.
2022-09-07T09:45:48.6210996Z Traceback of TorchScript (most recent call last):
2022-09-07T09:45:48.6211823Z RuntimeError: CUDA driver error: 200
```
and the assert seeing in the log is just checking on nvrtcCompileProgram being successful.
`2022-09-07T09:45:49.8770329Z RuntimeError: false INTERNAL ASSERT FAILED at "/var/lib/jenkins/workspace/torch/csrc/jit/codegen/cuda/executor_utils.cpp":1180, please report a bug to PyTorch. #pragma clang force_cuda_host_device begin
`

cc'ing @jeffdaily in case you want to take a look. Meanwhile, should we disable the tests for merge later when I resolve the macos build issues? @davidberard98

jjsjann123 commented:
wait... build CI on this PR actually passed. :tada: 

There's some cxx11-abi test that's failing, but looks like just permission issue, maybe it's my account?! https://github.com/pytorch/pytorch/runs/8233758234?check_suite_focus=true

davidberard98 commented:
I'm guessing the cxx11-abi failure is an infra failure, it looks like most PRs on master are also failing https://hud.pytorch.org/. Sanity-checks job will be skipped once we rebase this PR. But we do need to fix or disable rocm issues, I'm fine with disabling as long as @jeffdaily agrees.

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39324552).

jeffdaily commented:
There are a number of failed tests, but they all seem to originate from the same block of generated code:

```
/tmp/comgr-1caa3b/input/CompileSource:5601:5: error: function template partial specialization is not allowed
    reduce<Func, Types...>(
    ^     ~~~~~~~~~~~~~~~~
/tmp/comgr-1caa3b/input/CompileSource:6018:5: error: function template partial specialization is not allowed
    reduce<Func, Types...>(
    ^     ~~~~~~~~~~~~~~~~
/tmp/comgr-1caa3b/input/CompileSource:6075:5: error: function template partial specialization is not allowed
    reduceGroup<DataTypes..., Funcs..., BoolTypes...>(
    ^          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/comgr-1caa3b/input/CompileSource:6260:5: error: function template partial specialization is not allowed
    reduceGroup<DataTypes..., Funcs..., BoolTypes...>(
    ^          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/comgr-1caa3b/input/CompileSource:6320:5: error: function template partial specialization is not allowed
    reduceGroupBlock<BLOCK_BROADCAST, DataTypes..., Funcs..., BoolTypes...>(
    ^               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/comgr-1caa3b/input/CompileSource:6397:5: error: function template partial specialization is not allowed
    reduceGroupLastBlock<DataTypes..., Funcs..., BoolTypes...>(
    ^                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/comgr-1caa3b/input/CompileSource:6757:5: error: function template partial specialization is not allowed
    welfordGroup<NumArgs, DataType, IndexType>(
    ^           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/comgr-1caa3b/input/CompileSource:6958:5: error: function template partial specialization is not allowed
    welfordGroupBlock<BLOCK_BROADCAST, NumVals, DataType, IndexType>(
    ^                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/comgr-1caa3b/input/CompileSource:7027:5: error: function template partial specialization is not allowed
    welfordGroupLastBlock<NumVals, DataType, IndexType>(
    ^                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

jjsjann123 commented:
> /tmp/comgr-1caa3b/input/CompileSource:6075:5: error: function template partial specialization is not allowed
    reduceGroup<DataTypes..., Funcs..., BoolTypes...>(
    ^          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/comgr-1caa3b/input/CompileSource:6260:5: error: function template partial specialization is not allowed
    reduceGroup<DataTypes..., Funcs..., BoolTypes...>(


But reduceGroup here is not being partially specialized... it's just an overload and we do have all template arguments specified.
Sounds like a rocm compiler support thing. We can't back out cuda support because rocm is broken. Any chance we have a quick WAR this @jeffdaily ?

cc'ing @naoyam in case I was wrong about the overloaded template functions.

naoyam commented:
I'm not aware of any code that results in partial specialization of function templates. Where can I see generated code?

jjsjann123 commented:
> I'm not aware of any code that results in partial specialization of function templates. Where can I see generated code?

Thanks for confirming this.
This is on rocm stack and I assume you'll need AMD hardware, which is why we really can't provide much help from NV side on rocm failures... We'll see if @jeffdaily would be able to provide a quick patch. We can also turn off rocm tests and handle the failure in parallel. So we are at least not blocked on this.

jeffdaily commented:
You don't need AMD hardware for build errors.

docker pull rocm/dev-ubuntu-20.04:5.2

[rocm_build_error.cu.txt](https://github.com/pytorch/pytorch/files/9522207/rocm_build_error.cu.txt)

(github doesn't support uploading *.cu files, so *.txt extension added to file)

Inside docker: 

`hipcc -c rocm_build_error.cu -std=c++17  --amdgpu-target=gfx906`

davidberard98 commented:
@pytorchbot rebase -s

jjsjann123 committed:
Update on "[NVFuser] Upstream push 0907"


Syncing nvfuser devel branch to upstream master. https://github.com/csarofeen/pytorch/

Codegen changes include:

- codegen improvement:
i. improved view support on pointwise and transpose scheduler
ii. grouped grid welford added for better outer-norm grid persistence in normalization

- misc:
i. new composite ops added: variance_mean , arange, 
ii. fixes misaligned address for transpose scheduler
iii. refactor on separation of compilation API from execution API to prepare us for async compilation
iv. double type support on expression evaluator
v. PYTORCH_NVFUSER_DUMP refactor to save PTX and CUBIN

Commits that's in this PR from the devel branch:
```
89330aa23aa804340b2406ab58899d816e3dc3d2 Tensor factories must set the output shape as its input (#1939)
b2fd01ea9346712c6d6f623ca6addbc4888d008e arange support (#1933)
56c00fd3922dad7dfc57351ad7d780f0f2f8e4ed Double support on all expression evaluators (#1937)
371f28223e57fe3f6b5e50a0a45177e6a5c0785c Improve trivial reduction merge support (#1931)
1d0c26790e5647920b40d419d26815bbe310b3a6 Test `rand` in a fusion with zero tensor input (#1932)
0dab160fb2177d178eef3148c6a529e0855009e9 Fix softmax bwd sizes. (#1890)
ef98f360f6d3e3e1cc662ecb65202d88150f128d Fix a bug (#1936)
63132a0c56508c550084b07fb76a3df865102d00 Propagate permissive mapping information into indexing pass (#1929)
b4ac2c88d78078ee4d8b21c4fc51645b5710a282 Map IterationDomains through view operations. (#1919)
c0a187a7619d7cf9dc920294e15461791e8d6d4d do not use deprecated functions (#1935)
88de85e758c5e4afb7b6e746573c0d9a53b4cea7 Upstream cherry pick fixes 0811 (#1934)
b247dcf7c57dc6ac3f7a799b0a6beb7770536a74 Separate kernel compilation API from kernel execution API (#1914)
b34e3b93ee1a8030730c14af3995dd95665af07d Fix `ir_utils::hasBlockSync` + misc fixes in transpose scheduler (#1924)
14a53e6707f43bf760494c238a46386d69830822 Nullary RNGOp (#1892)
3c3c89e638f5172cafb0761f22bacd1fd695eec3 Misc fixes/tuning for transpose scheduler (#1912)
20cf109c8b44d48f61977e35bae94368985144ac Grouped grid welford (#1921)
6cf7eb024c9e53c358cbe56597e117bad56efefd Transpose scheduler small dim sizes better support (#1910)
9341ea9a5bf42f9b14ccad0c94edbc79fc5bb552 Disabled ViewPersistentShmoo sizes that results in NAN (#1922)
057237f66deeea816bb943d802a97c1b7e4414ab Fix CUDA driver error: misaligned address for transpose scheduler  (#1918)
3fb3d80339e4f794767a53eb8fdd61e64cf404a2 Add variance_mean function using Welford (#1907)
98febf6aa3b8c6fe4fdfb2864cda9e5d30089262 Remove DisableOption::UnrollWithRng (#1913)
ee8ef33a5591b534cf587d347af11e48ba7a15d4 Minor fix for the debug interface of using PTX directly (#1917)
6e8f953351f9dabfd1f991d8431cecb6c2ce684d Add PYTORCH_NVFUSER_DUMP options to save PTX and CUBIN (#1916)
5eefa9a72385f6a4b145680a9dcc52d7e8293763 dopt is only available since nvrtc 11.7 (#1915)
2ec8fc711eafc72451eebf0f5e2a98a38bf3f6ef Kill computeAtBetween (#1911)
d0d106a1d9af118d71673173674e875be35d259d Improve view support on pointwise and transpose scheduler (#1906)
e71e1ecefe67219846070590bbed54bbc7416b79 Fix name clash of RNG with shared memory (#1904)
3381793a253689abf224febc73fd3fe2a0dbc921 Fix mutator and sameAs for expanded IterDomain (#1902)
```

RUN_TORCHBENCH: nvfuser

Differential Revision: [D39324552](https://our.internmc.facebook.com/intern/diff/D39324552)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39324552).

jjsjann123 commented:
> You don't need AMD hardware for build errors.
> 
> docker pull rocm/dev-ubuntu-20.04:5.2
> 
> [rocm_build_error.cu.txt](https://github.com/pytorch/pytorch/files/9522207/rocm_build_error.cu.txt)
> 
> (github doesn't support uploading *.cu files, so *.txt extension added to file)
> 
> Inside docker:
> 
> `hipcc -c rocm_build_error.cu -std=c++17 --amdgpu-target=gfx906`
> 
> @naoyam the code is attached.

I don't think we have bandwidth to spare on rocm failures for our upstream PRs at the moment. This is not something we promised to support in nvfuser.
I'm suggesting that we disable the test for this PR. We'd be more than happy to review future PRs that patches and reenables related tests.

jeffdaily commented:
> I don't think we have bandwidth to spare on rocm failures for our upstream PRs at the moment. This is not something we promised to support in nvfuser.

At what point is ROCm truly going to be accepted as part of the pytorch community such that all PRs are expected to maintain the integrity of the ROCm builds and tests?  If anyone submitted a PR that broke some aspect of the CPU or CUDA builds, the procedure is to fix the broken PR and not skip tests.

I can understand not having bandwidth to fix a numerical issue that might be specific to ROCm because access to AMD GPUs can be a hurdle.  But the current issue is at compile-time, which _can_ be solved by anyone using the docker container and repro steps I have provided.

That said, we are dedicating resources to find a work-around for the current issue, but there is no ETA.

naoyam commented:
@jeffdaily Thanks for the generated code. Have you seen these errors before? As far as I can see, none of them is partial specialization, but I may be making some stupid template mistakes. Those errors seem to show up due to some of the recent changes I made so that the file sizes do not trigger some limits enforced in MSVC. Specifically, they were previously declared and defined in the class definition, but to make the file size smaller than the limit, I moved the function definitions out of the class.

naoyam commented:
This is the change that moved the functions definitions. https://github.com/csarofeen/pytorch/commit/81b1d36a1f755d478c1491d319d8c29a01680c7b

jeffdaily commented:
> @jeffdaily Thanks for the generated code. Have you seen these errors before? As far as I can see, none of them is partial specialization, but I may be making some stupid template mistakes. Those errors seem to show up due to some of the recent changes I made so that the file sizes do not trigger some limits enforced in MSVC. Specifically, they were previously declared and defined in the class definition, but to make the file size smaller than the limit, I moved the function definitions out of the class.

I haven't seen this specific error before.  I'm currently trying to write a smaller reproducer in hopes of finding a workaround.

jeffdaily commented:
Managed to repro for gcc 9.4.  Hopefully the repro.cpp faithfully recreates the sources.

[repro.cpp.txt](https://github.com/pytorch/pytorch/files/9529715/repro.cpp.txt)

gcc -c repro.cpp -std=c++14

```
repro.cpp:909:26: error: non-class, non-variable partial specialization ‘reduce<Func, Types ...>’ is not allowed
  909 |         Func reduction_op) {
      |                          ^
```

naoyam commented:
> > PRs that patches and reenables related tests.

Thanks for the repro. I see that for a template method of a template class, the type specification with `<>` should not be used for the function name, although it is necessary for the class name. I assume nvcc is just not very strict. That is my fault. Thanks so much for the repro. We'll fix the error.

jeffdaily commented:
I think I found a solution.  It worked for gcc and hip-clang.  

> > > PRs that patches and reenables related tests.
> 
> Thanks for the repro. I see that for a template method of a template class, the type specification with `<>` should not be used for the function name, although it is necessary for the class name. I assume nvcc is just not very strict. That is my fault. Thanks so much for the repro. We'll fix the error.

I was just about to send you the same suggestion.  I agree with your assessment.

jjsjann123 commented:
> I think I found a solution. It worked for gcc and hip-clang.
> 
> > > > PRs that patches and reenables related tests.
> > 
> > 
> > Thanks for the repro. I see that for a template method of a template class, the type specification with `<>` should not be used for the function name, although it is necessary for the class name. I assume nvcc is just not very strict. That is my fault. Thanks so much for the repro. We'll fix the error.
> 
> I was just about to send you the same suggestion. I agree with your assessment.

Thanks a lot for the debug and the repro. Naoya is patching the faulty templates and we are also following up with nvcc not catching the error on an internal bug.

I'll update the PR once we get the patch ready and hopefully everything will go through this time :crossed_fingers:

jjsjann123 committed:
Update on "[NVFuser] Upstream push 0907"


Syncing nvfuser devel branch to upstream master. https://github.com/csarofeen/pytorch/

Codegen changes include:

- codegen improvement:
i. improved view support on pointwise and transpose scheduler
ii. grouped grid welford added for better outer-norm grid persistence in normalization

- misc:
i. new composite ops added: variance_mean , arange, 
ii. fixes misaligned address for transpose scheduler
iii. refactor on separation of compilation API from execution API to prepare us for async compilation
iv. double type support on expression evaluator
v. PYTORCH_NVFUSER_DUMP refactor to save PTX and CUBIN

Commits that's in this PR from the devel branch:
```
89330aa23aa804340b2406ab58899d816e3dc3d2 Tensor factories must set the output shape as its input (#1939)
b2fd01ea9346712c6d6f623ca6addbc4888d008e arange support (#1933)
56c00fd3922dad7dfc57351ad7d780f0f2f8e4ed Double support on all expression evaluators (#1937)
371f28223e57fe3f6b5e50a0a45177e6a5c0785c Improve trivial reduction merge support (#1931)
1d0c26790e5647920b40d419d26815bbe310b3a6 Test `rand` in a fusion with zero tensor input (#1932)
0dab160fb2177d178eef3148c6a529e0855009e9 Fix softmax bwd sizes. (#1890)
ef98f360f6d3e3e1cc662ecb65202d88150f128d Fix a bug (#1936)
63132a0c56508c550084b07fb76a3df865102d00 Propagate permissive mapping information into indexing pass (#1929)
b4ac2c88d78078ee4d8b21c4fc51645b5710a282 Map IterationDomains through view operations. (#1919)
c0a187a7619d7cf9dc920294e15461791e8d6d4d do not use deprecated functions (#1935)
88de85e758c5e4afb7b6e746573c0d9a53b4cea7 Upstream cherry pick fixes 0811 (#1934)
b247dcf7c57dc6ac3f7a799b0a6beb7770536a74 Separate kernel compilation API from kernel execution API (#1914)
b34e3b93ee1a8030730c14af3995dd95665af07d Fix `ir_utils::hasBlockSync` + misc fixes in transpose scheduler (#1924)
14a53e6707f43bf760494c238a46386d69830822 Nullary RNGOp (#1892)
3c3c89e638f5172cafb0761f22bacd1fd695eec3 Misc fixes/tuning for transpose scheduler (#1912)
20cf109c8b44d48f61977e35bae94368985144ac Grouped grid welford (#1921)
6cf7eb024c9e53c358cbe56597e117bad56efefd Transpose scheduler small dim sizes better support (#1910)
9341ea9a5bf42f9b14ccad0c94edbc79fc5bb552 Disabled ViewPersistentShmoo sizes that results in NAN (#1922)
057237f66deeea816bb943d802a97c1b7e4414ab Fix CUDA driver error: misaligned address for transpose scheduler  (#1918)
3fb3d80339e4f794767a53eb8fdd61e64cf404a2 Add variance_mean function using Welford (#1907)
98febf6aa3b8c6fe4fdfb2864cda9e5d30089262 Remove DisableOption::UnrollWithRng (#1913)
ee8ef33a5591b534cf587d347af11e48ba7a15d4 Minor fix for the debug interface of using PTX directly (#1917)
6e8f953351f9dabfd1f991d8431cecb6c2ce684d Add PYTORCH_NVFUSER_DUMP options to save PTX and CUBIN (#1916)
5eefa9a72385f6a4b145680a9dcc52d7e8293763 dopt is only available since nvrtc 11.7 (#1915)
2ec8fc711eafc72451eebf0f5e2a98a38bf3f6ef Kill computeAtBetween (#1911)
d0d106a1d9af118d71673173674e875be35d259d Improve view support on pointwise and transpose scheduler (#1906)
e71e1ecefe67219846070590bbed54bbc7416b79 Fix name clash of RNG with shared memory (#1904)
3381793a253689abf224febc73fd3fe2a0dbc921 Fix mutator and sameAs for expanded IterDomain (#1902)
```

RUN_TORCHBENCH: nvfuser

Differential Revision: [D39324552](https://our.internmc.facebook.com/intern/diff/D39324552)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39324552).

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39324552).

jjsjann123 commented:
Public CI is all green~~ Victory~~~ :tada:

malfet commented on c10/util/hash.h:
```diff
@@ -304,6 +304,14 @@ struct hash<std::tuple<Types...>> {
+  size_t operator()(const std::pair<T1, T2>& pair) const {
+    std::tuple<T1, T2> tuple = std::make_tuple(pair.first, pair.second);
+    return _hash_detail::simple_get_hash(tuple);
+  };
```
Nit (internal clang-tidy complains about this one, not sure why OSS one does not)
```suggestion
  }
```

jjsjann123 replied:
:face_with_head_bandage: header change leads to rebuild of 1390 files

malfet commented on torch/csrc/jit/codegen/cuda/fusion.h:
```diff
@@ -4,6 +4,7 @@
 #include <c10/macros/Export.h>
 #include <c10/util/Exception.h>
 
+#include <torch/csrc/jit/codegen/cuda/executor_kernel_arg.h>
```
```suggestion
```

malfet commented on torch/csrc/jit/codegen/cuda/fusion.h:
```diff
@@ -168,18 +169,19 @@ class TORCH_CUDA_CU_API Fusion : public IrContainer {
   //! Run fusion segmentation algorithm to create a segmented fusion
-  std::unique_ptr<SegmentedFusion> segment(
-      const at::ArrayRef<at::IValue>& inputs);
+  std::unique_ptr<SegmentedFusion> segment(const KernelArgumentHolder& args);
```
```suggestion
  class KernelArgumentHolder;
  
  std::unique_ptr<SegmentedFusion> segment(const KernelArgumentHolder& args);
```

malfet reviewed (commented):
Proposed changes fixes internal builds at least for one config, will know the results for more later

jjsjann123 commented:
> Proposed changes fixes internal builds at least for one config, will know the results for more later

Thanks a lot for looking into this. I'm not sure if ghstack would take github suggested changes nicely, so I'm gonna play it safe and rebase locally. :fearful:

jjsjann123 committed:
Update on "[NVFuser] Upstream push 0907"


Syncing nvfuser devel branch to upstream master. https://github.com/csarofeen/pytorch/

Codegen changes include:

- codegen improvement:
i. improved view support on pointwise and transpose scheduler
ii. grouped grid welford added for better outer-norm grid persistence in normalization

- misc:
i. new composite ops added: variance_mean , arange, 
ii. fixes misaligned address for transpose scheduler
iii. refactor on separation of compilation API from execution API to prepare us for async compilation
iv. double type support on expression evaluator
v. PYTORCH_NVFUSER_DUMP refactor to save PTX and CUBIN

Commits that's in this PR from the devel branch:
```
89330aa23aa804340b2406ab58899d816e3dc3d2 Tensor factories must set the output shape as its input (#1939)
b2fd01ea9346712c6d6f623ca6addbc4888d008e arange support (#1933)
56c00fd3922dad7dfc57351ad7d780f0f2f8e4ed Double support on all expression evaluators (#1937)
371f28223e57fe3f6b5e50a0a45177e6a5c0785c Improve trivial reduction merge support (#1931)
1d0c26790e5647920b40d419d26815bbe310b3a6 Test `rand` in a fusion with zero tensor input (#1932)
0dab160fb2177d178eef3148c6a529e0855009e9 Fix softmax bwd sizes. (#1890)
ef98f360f6d3e3e1cc662ecb65202d88150f128d Fix a bug (#1936)
63132a0c56508c550084b07fb76a3df865102d00 Propagate permissive mapping information into indexing pass (#1929)
b4ac2c88d78078ee4d8b21c4fc51645b5710a282 Map IterationDomains through view operations. (#1919)
c0a187a7619d7cf9dc920294e15461791e8d6d4d do not use deprecated functions (#1935)
88de85e758c5e4afb7b6e746573c0d9a53b4cea7 Upstream cherry pick fixes 0811 (#1934)
b247dcf7c57dc6ac3f7a799b0a6beb7770536a74 Separate kernel compilation API from kernel execution API (#1914)
b34e3b93ee1a8030730c14af3995dd95665af07d Fix `ir_utils::hasBlockSync` + misc fixes in transpose scheduler (#1924)
14a53e6707f43bf760494c238a46386d69830822 Nullary RNGOp (#1892)
3c3c89e638f5172cafb0761f22bacd1fd695eec3 Misc fixes/tuning for transpose scheduler (#1912)
20cf109c8b44d48f61977e35bae94368985144ac Grouped grid welford (#1921)
6cf7eb024c9e53c358cbe56597e117bad56efefd Transpose scheduler small dim sizes better support (#1910)
9341ea9a5bf42f9b14ccad0c94edbc79fc5bb552 Disabled ViewPersistentShmoo sizes that results in NAN (#1922)
057237f66deeea816bb943d802a97c1b7e4414ab Fix CUDA driver error: misaligned address for transpose scheduler  (#1918)
3fb3d80339e4f794767a53eb8fdd61e64cf404a2 Add variance_mean function using Welford (#1907)
98febf6aa3b8c6fe4fdfb2864cda9e5d30089262 Remove DisableOption::UnrollWithRng (#1913)
ee8ef33a5591b534cf587d347af11e48ba7a15d4 Minor fix for the debug interface of using PTX directly (#1917)
6e8f953351f9dabfd1f991d8431cecb6c2ce684d Add PYTORCH_NVFUSER_DUMP options to save PTX and CUBIN (#1916)
5eefa9a72385f6a4b145680a9dcc52d7e8293763 dopt is only available since nvrtc 11.7 (#1915)
2ec8fc711eafc72451eebf0f5e2a98a38bf3f6ef Kill computeAtBetween (#1911)
d0d106a1d9af118d71673173674e875be35d259d Improve view support on pointwise and transpose scheduler (#1906)
e71e1ecefe67219846070590bbed54bbc7416b79 Fix name clash of RNG with shared memory (#1904)
3381793a253689abf224febc73fd3fe2a0dbc921 Fix mutator and sameAs for expanded IterDomain (#1902)
```

RUN_TORCHBENCH: nvfuser

Differential Revision: [D39324552](https://our.internmc.facebook.com/intern/diff/D39324552)

[ghstack-poisoned]

jjsjann123 committed:
Update on "[NVFuser] Upstream push 0907"


Syncing nvfuser devel branch to upstream master. https://github.com/csarofeen/pytorch/

Codegen changes include:

- codegen improvement:
i. improved view support on pointwise and transpose scheduler
ii. grouped grid welford added for better outer-norm grid persistence in normalization

- misc:
i. new composite ops added: variance_mean , arange, 
ii. fixes misaligned address for transpose scheduler
iii. refactor on separation of compilation API from execution API to prepare us for async compilation
iv. double type support on expression evaluator
v. PYTORCH_NVFUSER_DUMP refactor to save PTX and CUBIN

Commits that's in this PR from the devel branch:
```
89330aa23aa804340b2406ab58899d816e3dc3d2 Tensor factories must set the output shape as its input (#1939)
b2fd01ea9346712c6d6f623ca6addbc4888d008e arange support (#1933)
56c00fd3922dad7dfc57351ad7d780f0f2f8e4ed Double support on all expression evaluators (#1937)
371f28223e57fe3f6b5e50a0a45177e6a5c0785c Improve trivial reduction merge support (#1931)
1d0c26790e5647920b40d419d26815bbe310b3a6 Test `rand` in a fusion with zero tensor input (#1932)
0dab160fb2177d178eef3148c6a529e0855009e9 Fix softmax bwd sizes. (#1890)
ef98f360f6d3e3e1cc662ecb65202d88150f128d Fix a bug (#1936)
63132a0c56508c550084b07fb76a3df865102d00 Propagate permissive mapping information into indexing pass (#1929)
b4ac2c88d78078ee4d8b21c4fc51645b5710a282 Map IterationDomains through view operations. (#1919)
c0a187a7619d7cf9dc920294e15461791e8d6d4d do not use deprecated functions (#1935)
88de85e758c5e4afb7b6e746573c0d9a53b4cea7 Upstream cherry pick fixes 0811 (#1934)
b247dcf7c57dc6ac3f7a799b0a6beb7770536a74 Separate kernel compilation API from kernel execution API (#1914)
b34e3b93ee1a8030730c14af3995dd95665af07d Fix `ir_utils::hasBlockSync` + misc fixes in transpose scheduler (#1924)
14a53e6707f43bf760494c238a46386d69830822 Nullary RNGOp (#1892)
3c3c89e638f5172cafb0761f22bacd1fd695eec3 Misc fixes/tuning for transpose scheduler (#1912)
20cf109c8b44d48f61977e35bae94368985144ac Grouped grid welford (#1921)
6cf7eb024c9e53c358cbe56597e117bad56efefd Transpose scheduler small dim sizes better support (#1910)
9341ea9a5bf42f9b14ccad0c94edbc79fc5bb552 Disabled ViewPersistentShmoo sizes that results in NAN (#1922)
057237f66deeea816bb943d802a97c1b7e4414ab Fix CUDA driver error: misaligned address for transpose scheduler  (#1918)
3fb3d80339e4f794767a53eb8fdd61e64cf404a2 Add variance_mean function using Welford (#1907)
98febf6aa3b8c6fe4fdfb2864cda9e5d30089262 Remove DisableOption::UnrollWithRng (#1913)
ee8ef33a5591b534cf587d347af11e48ba7a15d4 Minor fix for the debug interface of using PTX directly (#1917)
6e8f953351f9dabfd1f991d8431cecb6c2ce684d Add PYTORCH_NVFUSER_DUMP options to save PTX and CUBIN (#1916)
5eefa9a72385f6a4b145680a9dcc52d7e8293763 dopt is only available since nvrtc 11.7 (#1915)
2ec8fc711eafc72451eebf0f5e2a98a38bf3f6ef Kill computeAtBetween (#1911)
d0d106a1d9af118d71673173674e875be35d259d Improve view support on pointwise and transpose scheduler (#1906)
e71e1ecefe67219846070590bbed54bbc7416b79 Fix name clash of RNG with shared memory (#1904)
3381793a253689abf224febc73fd3fe2a0dbc921 Fix mutator and sameAs for expanded IterDomain (#1902)
```

RUN_TORCHBENCH: nvfuser

Differential Revision: [D39324552](https://our.internmc.facebook.com/intern/diff/D39324552)

[ghstack-poisoned]

jjsjann123 commented on torch/csrc/jit/codegen/cuda/fusion.h:
```diff
@@ -4,6 +4,7 @@
 #include <c10/macros/Export.h>
 #include <c10/util/Exception.h>
 
+#include <torch/csrc/jit/codegen/cuda/executor_kernel_arg.h>
```
hmm. quite surprised how this didn't show up on public cpu build.

malfet commented:
> Thanks a lot for looking into this. I'm not sure if ghstack would take github suggested changes nicely, so I'm gonna play it safe and rebase locally. 😨

Stack will not pick suggested changes, so thank you for applying. After fwd declaring `KernelArgumentHolder` one would need to add explicit `#include <torch/csrc/jit/codegen/cuda/executor_kernel_arg.h>` to `torch/csrc/jit/codegen/cuda/evaluator_common.h` (I guess you've already done it)

jjsjann123 commented:
> > Thanks a lot for looking into this. I'm not sure if ghstack would take github suggested changes nicely, so I'm gonna play it safe and rebase locally. fearful
> 
> Stack will not pick suggested changes, so thank you for applying. After fwd declaring `KernelArgumentHolder` one would need to add explicit `#include <torch/csrc/jit/codegen/cuda/executor_kernel_arg.h>` to `torch/csrc/jit/codegen/cuda/evaluator_common.h` (I guess you've already done it)

Thanks for pointing it out~
That has already been done. I did a local build, which luckily caught that as well.

jjsjann123 commented:
@davidberard98 Can we import again to check internal build with the fixes. :bow:

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39324552).

jjsjann123 commented:
`The Diff and Pull Request are not in sync!` does this mean I need to rebase this PR?

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39324552).

davidberard98 commented:
> `Meta Internal-Only Changes Check The Diff and Pull Request are not in sync!` does this mean I need to rebase this PR?

sorry missed this last week, it usually means I need to re-import the diff.

jjsjann123 commented:
> > `Meta Internal-Only Changes Check The Diff and Pull Request are not in sync!` does this mean I need to rebase this PR?
> 
> sorry missed this last week, it usually means I need to re-import the diff.

No worries~ I'm on vacation this week, so won't be online during day time. But feel free to leave messages in this PR on what's needed. I'll try to address them when I get internet.

malfet commented:
Both OSS and internal CI are green, started merge process

davidberard98 commented:
@jjsjann123 when you have a chance can you rebase? unfortunately there have been a lot of internal failures when I'm trying to land (I think it's all unrelated failures) and by now there seem to be some merge conflicts.

jjsjann123 committed:
Update on "[NVFuser] Upstream push 0907"


Syncing nvfuser devel branch to upstream master. https://github.com/csarofeen/pytorch/

Codegen changes include:

- codegen improvement:
i. improved view support on pointwise and transpose scheduler
ii. grouped grid welford added for better outer-norm grid persistence in normalization

- misc:
i. new composite ops added: variance_mean , arange, 
ii. fixes misaligned address for transpose scheduler
iii. refactor on separation of compilation API from execution API to prepare us for async compilation
iv. double type support on expression evaluator
v. PYTORCH_NVFUSER_DUMP refactor to save PTX and CUBIN

Commits that's in this PR from the devel branch:
```
89330aa23aa804340b2406ab58899d816e3dc3d2 Tensor factories must set the output shape as its input (#1939)
b2fd01ea9346712c6d6f623ca6addbc4888d008e arange support (#1933)
56c00fd3922dad7dfc57351ad7d780f0f2f8e4ed Double support on all expression evaluators (#1937)
371f28223e57fe3f6b5e50a0a45177e6a5c0785c Improve trivial reduction merge support (#1931)
1d0c26790e5647920b40d419d26815bbe310b3a6 Test `rand` in a fusion with zero tensor input (#1932)
0dab160fb2177d178eef3148c6a529e0855009e9 Fix softmax bwd sizes. (#1890)
ef98f360f6d3e3e1cc662ecb65202d88150f128d Fix a bug (#1936)
63132a0c56508c550084b07fb76a3df865102d00 Propagate permissive mapping information into indexing pass (#1929)
b4ac2c88d78078ee4d8b21c4fc51645b5710a282 Map IterationDomains through view operations. (#1919)
c0a187a7619d7cf9dc920294e15461791e8d6d4d do not use deprecated functions (#1935)
88de85e758c5e4afb7b6e746573c0d9a53b4cea7 Upstream cherry pick fixes 0811 (#1934)
b247dcf7c57dc6ac3f7a799b0a6beb7770536a74 Separate kernel compilation API from kernel execution API (#1914)
b34e3b93ee1a8030730c14af3995dd95665af07d Fix `ir_utils::hasBlockSync` + misc fixes in transpose scheduler (#1924)
14a53e6707f43bf760494c238a46386d69830822 Nullary RNGOp (#1892)
3c3c89e638f5172cafb0761f22bacd1fd695eec3 Misc fixes/tuning for transpose scheduler (#1912)
20cf109c8b44d48f61977e35bae94368985144ac Grouped grid welford (#1921)
6cf7eb024c9e53c358cbe56597e117bad56efefd Transpose scheduler small dim sizes better support (#1910)
9341ea9a5bf42f9b14ccad0c94edbc79fc5bb552 Disabled ViewPersistentShmoo sizes that results in NAN (#1922)
057237f66deeea816bb943d802a97c1b7e4414ab Fix CUDA driver error: misaligned address for transpose scheduler  (#1918)
3fb3d80339e4f794767a53eb8fdd61e64cf404a2 Add variance_mean function using Welford (#1907)
98febf6aa3b8c6fe4fdfb2864cda9e5d30089262 Remove DisableOption::UnrollWithRng (#1913)
ee8ef33a5591b534cf587d347af11e48ba7a15d4 Minor fix for the debug interface of using PTX directly (#1917)
6e8f953351f9dabfd1f991d8431cecb6c2ce684d Add PYTORCH_NVFUSER_DUMP options to save PTX and CUBIN (#1916)
5eefa9a72385f6a4b145680a9dcc52d7e8293763 dopt is only available since nvrtc 11.7 (#1915)
2ec8fc711eafc72451eebf0f5e2a98a38bf3f6ef Kill computeAtBetween (#1911)
d0d106a1d9af118d71673173674e875be35d259d Improve view support on pointwise and transpose scheduler (#1906)
e71e1ecefe67219846070590bbed54bbc7416b79 Fix name clash of RNG with shared memory (#1904)
3381793a253689abf224febc73fd3fe2a0dbc921 Fix mutator and sameAs for expanded IterDomain (#1902)
```

RUN_TORCHBENCH: nvfuser

Differential Revision: [D39324552](https://our.internmc.facebook.com/intern/diff/D39324552)

[ghstack-poisoned]

jjsjann123 commented:
Sorry for the late response. Just bumped to latest `viable/strict`. :crossed_fingers:

jjsjann123 commented:
Oops. I'm actually seeing some errors on my local build. Taking a look.

jjsjann123 commented:
Looks like I missed the few lines of changes from here: https://github.com/pytorch/pytorch/pull/84967/files#diff-7a9ec1ab6cfe38cfdae3fe606acaca778a163c3015d19aa81adaffa0f38adcb3

jjsjann123 committed:
Update on "[NVFuser] Upstream push 0907"


Syncing nvfuser devel branch to upstream master. https://github.com/csarofeen/pytorch/

Codegen changes include:

- codegen improvement:
i. improved view support on pointwise and transpose scheduler
ii. grouped grid welford added for better outer-norm grid persistence in normalization

- misc:
i. new composite ops added: variance_mean , arange, 
ii. fixes misaligned address for transpose scheduler
iii. refactor on separation of compilation API from execution API to prepare us for async compilation
iv. double type support on expression evaluator
v. PYTORCH_NVFUSER_DUMP refactor to save PTX and CUBIN

Commits that's in this PR from the devel branch:
```
89330aa23aa804340b2406ab58899d816e3dc3d2 Tensor factories must set the output shape as its input (#1939)
b2fd01ea9346712c6d6f623ca6addbc4888d008e arange support (#1933)
56c00fd3922dad7dfc57351ad7d780f0f2f8e4ed Double support on all expression evaluators (#1937)
371f28223e57fe3f6b5e50a0a45177e6a5c0785c Improve trivial reduction merge support (#1931)
1d0c26790e5647920b40d419d26815bbe310b3a6 Test `rand` in a fusion with zero tensor input (#1932)
0dab160fb2177d178eef3148c6a529e0855009e9 Fix softmax bwd sizes. (#1890)
ef98f360f6d3e3e1cc662ecb65202d88150f128d Fix a bug (#1936)
63132a0c56508c550084b07fb76a3df865102d00 Propagate permissive mapping information into indexing pass (#1929)
b4ac2c88d78078ee4d8b21c4fc51645b5710a282 Map IterationDomains through view operations. (#1919)
c0a187a7619d7cf9dc920294e15461791e8d6d4d do not use deprecated functions (#1935)
88de85e758c5e4afb7b6e746573c0d9a53b4cea7 Upstream cherry pick fixes 0811 (#1934)
b247dcf7c57dc6ac3f7a799b0a6beb7770536a74 Separate kernel compilation API from kernel execution API (#1914)
b34e3b93ee1a8030730c14af3995dd95665af07d Fix `ir_utils::hasBlockSync` + misc fixes in transpose scheduler (#1924)
14a53e6707f43bf760494c238a46386d69830822 Nullary RNGOp (#1892)
3c3c89e638f5172cafb0761f22bacd1fd695eec3 Misc fixes/tuning for transpose scheduler (#1912)
20cf109c8b44d48f61977e35bae94368985144ac Grouped grid welford (#1921)
6cf7eb024c9e53c358cbe56597e117bad56efefd Transpose scheduler small dim sizes better support (#1910)
9341ea9a5bf42f9b14ccad0c94edbc79fc5bb552 Disabled ViewPersistentShmoo sizes that results in NAN (#1922)
057237f66deeea816bb943d802a97c1b7e4414ab Fix CUDA driver error: misaligned address for transpose scheduler  (#1918)
3fb3d80339e4f794767a53eb8fdd61e64cf404a2 Add variance_mean function using Welford (#1907)
98febf6aa3b8c6fe4fdfb2864cda9e5d30089262 Remove DisableOption::UnrollWithRng (#1913)
ee8ef33a5591b534cf587d347af11e48ba7a15d4 Minor fix for the debug interface of using PTX directly (#1917)
6e8f953351f9dabfd1f991d8431cecb6c2ce684d Add PYTORCH_NVFUSER_DUMP options to save PTX and CUBIN (#1916)
5eefa9a72385f6a4b145680a9dcc52d7e8293763 dopt is only available since nvrtc 11.7 (#1915)
2ec8fc711eafc72451eebf0f5e2a98a38bf3f6ef Kill computeAtBetween (#1911)
d0d106a1d9af118d71673173674e875be35d259d Improve view support on pointwise and transpose scheduler (#1906)
e71e1ecefe67219846070590bbed54bbc7416b79 Fix name clash of RNG with shared memory (#1904)
3381793a253689abf224febc73fd3fe2a0dbc921 Fix mutator and sameAs for expanded IterDomain (#1902)
```

RUN_TORCHBENCH: nvfuser

Differential Revision: [D39324552](https://our.internmc.facebook.com/intern/diff/D39324552)

[ghstack-poisoned]

jjsjann123 commented:
hmmm... upstream tests seems to be broken. I tried it on our nightlies without changes from this PR. Looks like running `PyFusionCache` after `FusionDefinition` gives the error. With a standalone `PyFusionCache` test, everything seems to be fine.

cc'ing @kevinstephano 

```
root@44382fd5d207:/opt/pytorch/pytorch# ./build/bin/test_jit --gtest_filter="*NVFuser*"
Note: Google Test filter = *NVFuser*
[==========] Running 688 tests from 2 test suites.
[----------] Global test environment set-up.
[----------] 686 tests from NVFuserTest
[ RUN      ] NVFuserTest.FusionDefinition_CUDA
[       OK ] NVFuserTest.FusionDefinition_CUDA (6 ms)
[ RUN      ] NVFuserTest.PyFusionCache_CUDA
/opt/pytorch/pytorch/torch/csrc/jit/codegen/cuda/python_frontend/test/test_nvfuser_fusion_cache.cpp:73: Failure
Failed
Unexpected assert during cache lookup!There should be no children from a Terminal Cache Entry!
Exception raised from lookupFusionCacheEntry at /opt/pytorch/pytorch/torch/csrc/jit/codegen/cuda/python_frontend/fusion_cache.cpp:63 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x6c (0x7f29d02588dc in /opt/pytorch/pytorch/build/lib/libc10.so)
frame #1: c10::detail::torchCheckFail(char const*, char const*, unsigned int, char const*) + 0xf5 (0x7f29d0224d77 in /opt/pytorch/pytorch/build/lib/libc10.so)
frame #2: nvfuser::FusionCache::lookupFusionCacheEntry(nvfuser::RecordFunctor*) const + 0x87 (0x7f29d36841d7 in /opt/pytorch/pytorch/build/lib/libtorch_cuda.so)
frame #3: <unknown function> + 0x27e50e (0x5564607bc50e in ./build/bin/test_jit)
frame #4: void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) + 0x51 (0x556460d3c771 in ./build/bin/test_jit)
frame #5: <unknown function> + 0x7f2110 (0x556460d30110 in ./build/bin/test_jit)
frame #6: <unknown function> + 0x7f25a5 (0x556460d305a5 in ./build/bin/test_jit)
frame #7: <unknown function> + 0x7f2d01 (0x556460d30d01 in ./build/bin/test_jit)
frame #8: testing::internal::UnitTestImpl::RunAllTests() + 0x10e9 (0x556460d32489 in ./build/bin/test_jit)
frame #9: testing::UnitTest::Run() + 0x98 (0x556460d329a8 in ./build/bin/test_jit)
frame #10: main + 0xcc (0x556460813fac in ./build/bin/test_jit)
frame #11: __libc_start_main + 0xf3 (0x7f29cfd62083 in /usr/lib/x86_64-linux-gnu/libc.so.6)
frame #12: _start + 0x2e (0x55646086780e in ./build/bin/test_jit)

[  FAILED  ] NVFuserTest.PyFusionCache_CUDA (6 ms)

```

jjsjann123 commented:
Tests looks clean locally, except the one I mentioned above. But that is not broken in this PR and doesn't seem to get caught by upstream CI. :confused: 
Anyway, I think we can try another import here now.

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39324552).

jjsjann123 commented:
Wondering if the bump has helped with internal CI failures. :eyes:

davidberard98 commented:
@jjsjann123 sorry, can we get one more rebase?

seems like the test runner for the internal tests failed the last few times, not sure why

davidberard98 commented:
@jjsjann123 sorry, can we get one more rebase?

seems like the test runner for the internal tests failed the last few times, not sure why

jeffdaily commented:
@davidberard98 any specifics for internal test failures?  ROCm-specific or no?

jjsjann123 commented:
> @davidberard98 any specifics for internal test failures? ROCm-specific or no?

Do we also have interla ROCm-specific tests? Asking since I don't have visibility in that and looks like public ROCm CIs are green here.

davidberard98 commented:
tests are just flaky from what I can tell. I'm not familiar with rocm tests internally.

jjsjann123 committed:
Update on "[NVFuser] Upstream push 0907"


Syncing nvfuser devel branch to upstream master. https://github.com/csarofeen/pytorch/

Codegen changes include:

- codegen improvement:
i. improved view support on pointwise and transpose scheduler
ii. grouped grid welford added for better outer-norm grid persistence in normalization

- misc:
i. new composite ops added: variance_mean , arange, 
ii. fixes misaligned address for transpose scheduler
iii. refactor on separation of compilation API from execution API to prepare us for async compilation
iv. double type support on expression evaluator
v. PYTORCH_NVFUSER_DUMP refactor to save PTX and CUBIN

Commits that's in this PR from the devel branch:
```
89330aa23aa804340b2406ab58899d816e3dc3d2 Tensor factories must set the output shape as its input (#1939)
b2fd01ea9346712c6d6f623ca6addbc4888d008e arange support (#1933)
56c00fd3922dad7dfc57351ad7d780f0f2f8e4ed Double support on all expression evaluators (#1937)
371f28223e57fe3f6b5e50a0a45177e6a5c0785c Improve trivial reduction merge support (#1931)
1d0c26790e5647920b40d419d26815bbe310b3a6 Test `rand` in a fusion with zero tensor input (#1932)
0dab160fb2177d178eef3148c6a529e0855009e9 Fix softmax bwd sizes. (#1890)
ef98f360f6d3e3e1cc662ecb65202d88150f128d Fix a bug (#1936)
63132a0c56508c550084b07fb76a3df865102d00 Propagate permissive mapping information into indexing pass (#1929)
b4ac2c88d78078ee4d8b21c4fc51645b5710a282 Map IterationDomains through view operations. (#1919)
c0a187a7619d7cf9dc920294e15461791e8d6d4d do not use deprecated functions (#1935)
88de85e758c5e4afb7b6e746573c0d9a53b4cea7 Upstream cherry pick fixes 0811 (#1934)
b247dcf7c57dc6ac3f7a799b0a6beb7770536a74 Separate kernel compilation API from kernel execution API (#1914)
b34e3b93ee1a8030730c14af3995dd95665af07d Fix `ir_utils::hasBlockSync` + misc fixes in transpose scheduler (#1924)
14a53e6707f43bf760494c238a46386d69830822 Nullary RNGOp (#1892)
3c3c89e638f5172cafb0761f22bacd1fd695eec3 Misc fixes/tuning for transpose scheduler (#1912)
20cf109c8b44d48f61977e35bae94368985144ac Grouped grid welford (#1921)
6cf7eb024c9e53c358cbe56597e117bad56efefd Transpose scheduler small dim sizes better support (#1910)
9341ea9a5bf42f9b14ccad0c94edbc79fc5bb552 Disabled ViewPersistentShmoo sizes that results in NAN (#1922)
057237f66deeea816bb943d802a97c1b7e4414ab Fix CUDA driver error: misaligned address for transpose scheduler  (#1918)
3fb3d80339e4f794767a53eb8fdd61e64cf404a2 Add variance_mean function using Welford (#1907)
98febf6aa3b8c6fe4fdfb2864cda9e5d30089262 Remove DisableOption::UnrollWithRng (#1913)
ee8ef33a5591b534cf587d347af11e48ba7a15d4 Minor fix for the debug interface of using PTX directly (#1917)
6e8f953351f9dabfd1f991d8431cecb6c2ce684d Add PYTORCH_NVFUSER_DUMP options to save PTX and CUBIN (#1916)
5eefa9a72385f6a4b145680a9dcc52d7e8293763 dopt is only available since nvrtc 11.7 (#1915)
2ec8fc711eafc72451eebf0f5e2a98a38bf3f6ef Kill computeAtBetween (#1911)
d0d106a1d9af118d71673173674e875be35d259d Improve view support on pointwise and transpose scheduler (#1906)
e71e1ecefe67219846070590bbed54bbc7416b79 Fix name clash of RNG with shared memory (#1904)
3381793a253689abf224febc73fd3fe2a0dbc921 Fix mutator and sameAs for expanded IterDomain (#1902)
```

RUN_TORCHBENCH: nvfuser

Differential Revision: [D39324552](https://our.internmc.facebook.com/intern/diff/D39324552)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39324552).

jjsjann123 commented:
Daily bump for update since I'm getting nervous as we approach deadline. :bow:

IvanYashchuk commented:
@malfet, @davidberard98 is there anything holding off this PR from being merged internally?

davidberard98 commented:
@pytorchbot merge

davidberard98 commented:
@jjsjann123 @IvanYashchuk I am merging manually because the internal diff has already been merged.  In general we still need to merge via the internal workflow; but in this case I am merging via github just because the bot hasn't done it for some reason.
